### DISCUSSION TITLE
SCTransform - Splitting Dataset

### DISCUSSION THREAD

**Discussion by NicoHadas** [satijalab/seurat#8441] [Q&A]
Hello,

I am working on a dataset with multiple samples that have different stimulated timepoints and I am trying to run SCTransform on each object/sample individually prior to performing integration. 

I was wondering about the use of the split function seen below that is used in [Introduction to scRNA-seq integration ](https://satijalab.org/seurat/articles/integration_introduction.html#perform-integration-with-sctransform-normalized-datasets) vignette. 
It is used prior to running SCTransform and CCA integration. 

dataset["RNA"]] <- split(dataset[["RNA"]], f = dataset$timepoint)

Does this function split the dataset, allowing one to run SCTransform on each layer individually. Or would I still have to run SCTransform on each sample separately and then merge ?

Any help/guidance is appreciated!

**Comment by haiderabbas678** [comment]
Yes I have been experiencing a similar issue in the work I was conducting. Hoping for a response

**Comment by alifarhat40** [comment]
Hi @NicoHadas did you figure this out? I was wondering the same for me.

**Reply by rachel7p** [reply]
see https://github.com/satijalab/seurat/discussions/8441#discussioncomment-12518585

**Comment by rachel7p** [comment]
Not a member of the dev team, but yes, 

dataset["RNA"]] <- split(dataset[["RNA"]], f = dataset$timepoint) 

allows one to run SCTransform on each layer individually. Then you run IntegrateLayers with normalization.method = "SCT".  
___________
As an aside: The above code is equivalent to the old way where you would SplitObject and iterate over each object like below.
dataset <- SplitObject(dataset, split.by = timepoint)
dataset <- lapply(dataset, SCTransform, vst.flavor = "v2",  verbose = FALSE)
___________

However, the problem most people are facing is that JoinLayers() does not seem to work on an SCTransform-ed object and you would need to use PrepSCTFindMarkers instead. This re-joins the layers but still lists multiple SCTmodels.  So, when running FindVariableFeatures, you get an error that says 'SCT assay is comprised of multiple SCT models. To change the variable features, please set manually with VariableFeatures<-.'

I saw a solution mentioned here https://github.com/satijalab/seurat/issues/5761#issuecomment-1079212702 to use SelectIntegrationFeatures, but that did not work for me. 

And then I found this https://github.com/satijalab/seurat/issues/4059 suggesting to re-run the PCA on the integrated assay and then try FindVariableFeatures. It didn't work until I also re-ran the SCTransform after PrepSCTFindMarkers. Here is my workflow:

# Split Layers by sample_id which stores my batch information
sc[["RNA"]] <- split(sc[["RNA"]], f = sc$sample_id)

#Run SCTransform
sc <- SCTransform(sc, vst.flavor = "v2",  verbose = FALSE)

sc <- RunPCA(sc, npcs = 30, features = VariableFeatures(sc))

# Integrate. I used a k.weight specific to my data.
sc <- IntegrateLayers(
  object = sc, method = CCAIntegration, assay = "SCT",
  orig.reduction = "pca", new.reduction = "integrated.cca", k.weight= 60,
  normalization.method = "SCT",  
  verbose = FALSE
)

# Find neighbors and clusters
sc <- FindNeighbors(sc, reduction = "integrated.cca", dims = 1:30)
sc <- FindClusters(sc, resolution = 2, cluster.name = "cca_clusters")

# Run UMAP
sc <- RunUMAP(sc, dims = 1:30, reduction = "integrated.cca", reduction.name= "umap.cca")

# Visualize the integration. I left this code out as it is not directly pertinent here.

# Re-join Layers. There are still multiple SCT.models in the object
sc <- PrepSCTFindMarkers(sc, assay = "SCT")

# Re-run SCTransform. This collapsed the multiples SCTmodels to just one.
sc <- SCTransform(sc, vst.flavor = "v2", verbose = FALSE)

# Run PCA
sc <- RunPCA(sc, npcs = 30, features = VariableFeatures(sc))

sc <-  FindVariableFeatures(sc, selection.method="vst", assay = "SCT", nfeatures=2000)

Continue with downstream analysis. I do not know if this is the correct way to do it, but it was the only way it would work for me after two days of searching vignettes and GitHub forums.

If anyone else has figured out a different approach, please chime in!

**Reply by alifarhat40** [reply]
@rachel7p if you look at this part of the tutorial here:
https://satijalab.org/seurat/archive/v4.3/sctransform_v2_vignette#:~:text=Perform%20integration%20using%20pearson%20residuals

It seems that they used functions such as 
ifnb.list <- [PrepSCTIntegration](https://satijalab.org/seurat/archive/reference/PrepSCTIntegration.html)(object.list = ifnb.list, anchor.features = features)

However, you did not use it in this code here. Is this an important step?
